src/java/voldemort/client/protocol/RequestFormat.java

```diff
@@ -68,6 +70,17 @@ public void writeGetAllRequest(DataOutputStream output,
     public Map<ByteArray, List<Versioned<byte[]>>> readGetAllResponse(DataInputStream stream)
             throws IOException;
 
+    /*
+     * returns the max expected put request size, An implementation can avoid it
+     * by returning the value SIZE_UNKNOWN
+     */
+    public int getExpectedPutRequestSize(String storeName,
```

Comment:
The Serve still has no clue whether it is pulling out a rabbit or elphant :) . That will require a protcol change. But currently the problem is let us say if you write a value that is 65KB. The client will allocate a ByteBuffer for 130KB ( if it exceeds the size, it doubles the buffer of the required size). And after that it shrinks it to 64KB again. With this fix, the client will just grow the buffer to 65KB and leave it there. I went after all the low hanging fruits , but the Protocol is left intact.